File: java/maximum-count-of-positive-integer-and-negative-integer.java
class Solution {
    public int maximumCount(int[] nums) {
        int len = nums.length;
        // using binary search to find the index of the first non-negative number i.e., l and that would be the answer
        int l = 0,r = len;
        while (l<r) {
            int mid = l + (r-l)/2;
            if (nums[mid]>=0) {
                r = mid;
            }
            else {
                l = mid+1;
            }
        }
        int neg = l;

        // using binary search to find the first index of positive number i.e., l and then len - l would be the answer
        l = 0;
        r = len;
        while (l<r) {
            int mid = l + (r-l)/2;
            if (nums[mid]>0) {
                r = mid;
            }
            else {
                l = mid+1;
            }
        }
        int pos = len - l;

        // max of these two would be the answer
        return Math.max(neg,pos);
    }
}

File: cpp/maximum-count-of-positive-integer-and-negative-integer.cpp
class Solution {
public:
    int maximumCount(vector<int>& nums) {
        // finding the position of the first positive number and then calculating number of positives using (size of vector) - (this position)
        int pos = lower_bound(nums.begin(),nums.end(),1)-nums.begin();
        // the number of negatives is the same as the index of the first non-negative number since non-decreasing vector
        int neg = lower_bound(nums.begin(),nums.end(),0)-nums.begin();
        pos = nums.size()-pos;
        return max(pos,neg);
    }
};
